content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md

```diff
@@ -0,0 +1,141 @@
+* ArchLinux and Fedora offer riscv64 toolchains.
+  * I recommend installing them. (ex: riscv64-elf-gdb, riscv64-elf-gcc, riscv64-elf-binutils)
+
+### Compiling Haiku (with debug symbols)
+
+Overall, this is the easier step for most folks with any Haiku development experience.
+
+1. Get the [haiku and buildtools git repos](/guides/building/get-source-git).
+2. Compile your toolchain.
+   1. ```cd haiku; mkdir generated.riscv64; cd generated.riscv64```
+   1. ```../configure -j2 --cross-tools-source ../../buildtools --build-cross-tools riscv64```
+3. Enable debugging symbols in bootloader.
+   1. Copy UserBuildConfig.example to UserBuildConfig in haiku/build/jam/
+   2. Add ```SetConfigVar DEBUG : HAIKU_TOP src system boot : 1 : global ;```
+4. Compile haiku for riscv64 ```jam -q @minimum-mmc```
+
+### Install GEF
+
+[GEF](https://github.com/hugsy/gef) is an extension for gdb making it a lot more friendly and will help visualize what's going on.
+
+### Compile u-boot
+
+u-boot is the early loader of choice for RISCV64 and ARM platforms.  You can use Haiku's firmware CI/CD tools to simplify getting and compiling u-boot.
+
+1. Clone Haiku's [firmware repo](https://github.com/haiku/firmware/)
+2. Build riscv64 qemu u-boot
+   1. ```cd firmware; cd u-boot; ./tools/build riscv64 qemu```
+
+After you have the firmware, we're less interested in the newly generated changes to the firmware repository, and more interested in /tmp/riscv64-build/ which contains all of the u-boot artifacts. (u-boot.bin for qemu, u-boot (elf, debugging sybmols) for GDB)
+
+## Get ready to debug
+
+Now we have all the componenents needed to begin debugging Haiku's bootloader.
+
+### Boot u-boot + Haiku in qemu-system-riscv64
+
+```bash
+qemu-system-riscv64 -bios /tmp/riscv64-build/u-boot.bin \
+  -M virt -m 4G \
+  -drive file=haiku-mmc.image,format=raw,id=hd0 -device virtio-blk-device,drive=hd0 \
+  -s -S
+```
+
+  * ```-bios ...```
+    * The bios is the u-boot bios you compiled
+  * ```-M virt```
+    * A standardized "qemu riscv64 machine"
+  * ```-m 4G```
+    * 4 GiB of memory
+  * ```-drive```
+    * Haiku's disk image (contains our EFI bootloader, and Haiku's filesystem)
+  * ```-s```
+    * Listen for GDB connections on 127.0.0.1:1234
+  * ```-S```
+    * Halt boot until GDB says to begin running
+  
+**At this point, qemu will "wait" for your command to start running.**
+
+### Attach GDB
+
+Now, we attach to qemu from GDB, and start debugging!
+
+*TIP: The hex addresses below could change over time. If you don't get debugging symbols (aka source-code view of what's happening) you'll need to "figure the new addresses out". I've included some basic information on how I got them.*
+
+1. Start GDB for your architecture
+   1. ```riscv64-elf-gdb /tmp/riscv64-build/u-boot``` on ArchLinux
+2. Connect to qemu
+   1. ```target remote 127.0.0.1:1234```
+3. Tell qemu where u-boot symbols should be loaded
```

Comment:
Actually it's "tell gdb", qemu does not care about symbols

Reply:
oops. nice catch!